cmake/Corrosion.cmake

```diff
@@ -587,10 +587,18 @@ function(corrosion_link_libraries target_name)
             PROPERTY CARGO_DEPS_LINKER_LANGUAGES
             $<TARGET_PROPERTY:${library},LINKER_LANGUAGE>
         )
-        corrosion_add_target_rustflags(${target_name} "-L$<TARGET_LINKER_FILE_DIR:${library}>")
 
-        # TODO: The output name of the library can be overridden - find a way to support that.
-        corrosion_add_target_rustflags(${target_name} "-l${library}")
+        get_property(
+            LIBRARY_OUTPUT_NAME
+            TARGET ${library}
+            PROPERTY OUTPUT_NAME
+        )
+        if (NOT LIBRARY_OUTPUT_NAME)
+            set(LIBRARY_OUTPUT_NAME ${library})
+        endif()
+
+        corrosion_add_target_rustflags(${target_name} "-L$<TARGET_LINKER_FILE_DIR:${library}>")
+        corrosion_add_target_rustflags(${target_name} "-l${LIBRARY_OUTPUT_NAME}")
```

Comment:
Couldn't we use [`$<TARGET_LINKER_FILE:tgt>`](https://cmake.org/cmake/help/latest/manual/cmake-generator-expressions.7.html#genex:TARGET_LINKER_FILE) here?

Looking at [LIBRARY_OUTPUT_NAME](https://cmake.org/cmake/help/latest/prop_tgt/LIBRARY_OUTPUT_NAME.html), it could be not set, but `OUTPUT_NAME` could be set. Using the `TARGET_LINKER_FILE` genex should probably automatically select the right version.

Reply:
hm. When I use that instead, it fails to link the test.

````
"-L" "/home/andrew/corrosion/build/cargo/build/release/deps" "-L" "/home/andrew/corrosion/build/test/cpp2rust" "-L" "/home/andrew/corrosion/build/test/cpp2rust" "-L" "/home/andrew/corrosion/build/test/cpp2rust" "-L" "/home/andrew/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-l/home/andrew/corrosion/build/test/cpp2rust/libcpp-lib.a" "-l/home/andrew/corrosion/build/test/cpp2rust/libcpp-lib-renamed.a" "-l/home/andrew/corrosion/build/test/cpp2rust/libcpp-lib3.a"  <SNIP> "-Wl,--gc-sections" "-pie" "-Wl,-zrelro,-znow" "-Wl,-O1"
  = note: /usr/bin/ld: cannot find -l/home/andrew/corrosion/build/test/cpp2rust/libcpp-lib.a
          /usr/bin/ld: cannot find -l/home/andrew/corrosion/build/test/cpp2rust/libcpp-lib-renamed.a
          /usr/bin/ld: cannot find -l/home/andrew/corrosion/build/test/cpp2rust/libcpp-lib3.a
          collect2: error: ld returned 1 exit status
````
but you're right, my solution does ignore LIBRARY_OUTPUT_NAME, ARCHIVE_OUTPUT_NAME and the `` _<CONFIG>`` variants, which needs fixing

Reply:
If CMake 3.15 is an ok base reqiurement, the TARGET_FILE_BASE_NAME genex could be used.

https://cmake.org/cmake/help/latest/manual/cmake-generator-expressions.7.html#genex:TARGET_FILE_BASE_NAME

Reply:
I think you need to add a `:` like `-l:/path/x.a`, see the [documenation](https://linux.die.net/man/1/ld) for `--library`. Not 100% sure this works though.

> If CMake 3.15 is an ok base reqiurement

In that case this either needs to be guarded by a version check, or wait until after Corrosion v0.2 is released (which I plan to do soon, I just have to update the readme to reflect the recent changes).

Reply:
I've poked around with this for about an hour now and can't figure out a proper way to do it that only requires the current minimum cmake of 3.12. putting `:` in the `-l` command screws up the INTERFACE_CORROSION_RUSTFLAGS generator expression and causes a funky compile error: 

e.g.
`
-- IFACE flags are: $<$<BOOL:$<$<NOT:$<BOOL:$<TARGET_PROPERTY:rust-exe,CORROSION_USE_HOST_BUILD>>>:>>:-Clink-args=$<$<NOT:$<BOOL:$<TARGET_PROPERTY:rust-exe,CORROSION_USE_HOST_BUILD>>>:>>;-Cdefault-linker-libraries=yes;-l:/home/andrew/corrosion/build/test/cpp2rust/libcpp-lib.a;-L$<TARGET_LINKER_FILE_DIR:cpp-lib2>;-l$<TARGET_LINKER_FILE_BASE_NAME:cpp-lib2>;-L$<TARGET_LINKER_FILE_DIR:cpp-lib3>;-l$<TARGET_LINKER_FILE_BASE_NAME:cpp-lib3>
`

```
error: renaming of the library `` was specified, however this crate contains no `#[link(...)]` attributes referencing this library
```

Similarly, trying to use the TARGET_LINKER_FILE genex but strip out the full path to just be the `get_filename_component(blah NAME_WE)` or such doesn't work, because the genex is expanded much later than configure time. 

So waiting for 0.2 to drop seems reasonable to me :) 

The symmetry of 
```
       corrosion_add_target_rustflags(${target_name} "-L$<TARGET_LINKER_FILE_DIR:${library}>")
        corrosion_add_target_rustflags(${target_name} "-l$<TARGET_LINKER_FILE_BASE_NAME:${library}>")
```
is pretty appealing at least.

Reply:
Thanks for taking the time to investigate! 

> So waiting for 0.2 to drop seems reasonable to me :)

Ok, great.

Reply:
I've pushed an update to use the preferred syntax, but given your announcement it will have to wait until v0.3 :)

Reply:
@ADKaster I've updated the minimum supported CMake version to 3.15. Could you rebase your changes?

Reply:
Sure thing, rebased on master :)